### PR TITLE
Split isExternal in lib/url

### PR DESCRIPTION
**Pull request by sgomes**
This is one of several PRs aiming at splitting up `lib/url` into several files for tree-shaking, replacing usage of the browserified node `url` module with native functionality, and generally improving it.

This PR focuses on `isExternal`, which I'm splitting into its own module and reimplementing using native functionality.

This PR was split out from #36531, which got too big.

#### Changes proposed in this Pull Request

* Split out `isExternal` in `lib/url`
* Split out corresponding tests into their own file too

#### Testing instructions

The unit tests for this method are pretty extensive, so they should be enough in ensuring that there are no undesirable behaviour changes.

### CODE REVIEW

**Comment by sgomes** [comment]
Rebased on `master`.

File: client/lib/url/is-external.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+export default function isExternal( url: TypedURL ): boolean {
+	// The url passed in might be of form `en.support.wordpress.com`,
+	// so for this function we'll append double-slashes to fake it.
+	// If it is a relative URL the hostname will be the base hostname.
+	if (
+		! url.startsWith( 'http://' ) &&
+		! url.startsWith( 'https://' ) &&
+		! url.startsWith( '/' ) &&
+		! url.startsWith( '?' ) &&
+		! url.startsWith( '#' )
```
Comment (sirreal) [inline client/lib/url/is-external.ts]: We're missing the case where the URL already starts with `//`, right?

```suggestion
		! url.startsWith( '#' ) &&
		! startsWith( url, '//' )
```

We don't want to modify the URL to be `////example.com`.

File: client/lib/url/test/is-external.js [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+ */
+import isExternal from '../is-external';
+
+describe( 'isExternal', () => {
+	test( 'should return false for relative path-only url', () => {
+		const source = '/relative';
+
+		const actual = isExternal( source );
+
+		expect( actual ).toBeFalse;
```
Comment (sirreal) [inline client/lib/url/test/is-external.js]: These look like chai expects, is this actually testing anything?

We can either rewrite them all for jest expect (preferred: `.toBe( false )`), or pull the import from the file they came from:

```js
import { expect } from 'chai';
```

**Review by sirreal — CHANGES_REQUESTED** [review]

File: client/lib/url/is-external.ts (hunk 1 above)
Comment (sgomes) [reply client/lib/url/is-external.ts]: Doesn't `url.startsWith( '/' )` already cover that case?

Comment (sirreal) [reply client/lib/url/is-external.ts]: 🤦‍♂

**Review by sirreal — APPROVED** [review]
Thanks! Looks good when tests + canaries pass 👍

File: client/lib/url/test/is-external.js [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+			expect( actual ).toBeFalsy();
+		} );
+
+		test( 'should return true for external protocol-relative url', () => {
+			const source = '//some-other-site.com';
+
+			const actual = isExternal( source );
+
+			expect( actual ).toBeTruthy();
```
Comment (sirreal) [inline client/lib/url/test/is-external.js]: When we want exactly `true`, we should prefer this form:

```suggestion
			expect( actual ).toBe( true );
```

Not a big deal 🙂

Comment (sgomes) [reply client/lib/url/test/is-external.js]: Yep, we can be stricter 🙂 Changed everything to `.toBe( true )`/`.toBe( false )`.

File: client/lib/url/is-external.ts (hunk 1 above)
Comment (sgomes) [reply client/lib/url/is-external.ts]: Haha, don't worry, it's easy to miss! 😄

**Comment by sgomes** [comment]
Thanks for the review, @sirreal!

**Pull request merged**